### PR TITLE
Fix IllegalStateException caused by Flows instrumentation in Kotlin coroutines

### PR DESCRIPTION
**Pull request by monosoul**
# What Does This Do
Makes `ScopeStateCoroutineContext` inheritable by introducing an internal map of a child coroutine (aka Job) (inheriting the context) to the context item holding the continuation. This way each child coroutine will hold the same context element while having individual state internally.
Solves #4718 .

# Motivation
To fix an `IllegalStateException` thrown by [this](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/common/src/flow/internal/SafeCollector.common.kt) Flows collector in case of a modified context.

# Additional Notes
This is a followup to #4405 and  #4624

### CODE REVIEW

**Comment by monosoul** [comment]
@bantonsson please, have a look. Thanks!

File: dd-java-agent/instrumentation/kotlin-coroutines/src/main/java/datadog/trace/instrumentation/kotlin/coroutines/ScopeStateCoroutineContext.java [hunk 1]
```diff
@@ -3,36 +3,27 @@
...
-  private final ScopeState coroutineScopeState;
-  @Nullable private AgentScope.Continuation continuation;
-  @Nullable private AgentScope continuationScope;
-  private boolean isInitialized = false;
 
-  public ScopeStateCoroutineContext() {
-    coroutineScopeState = AgentTracer.get().newScopeState();
-  }
+  private final Map<Job, ScopeStateCoroutineContextItem> contextItemPerCoroutine =
+      new ConcurrentHashMap<>();
```
Comment (bantonsson) [inline dd-java-agent/instrumentation/kotlin-coroutines/src/main/java/datadog/trace/instrumentation/kotlin/coroutines/ScopeStateCoroutineContext.java:20]: Instead of using a concurrent hash map, there is a helper called `ContextStore<K, C>` and `InstrumentationContext`, that allows you to store instances of `C` directly on the `K`. Simple example usage [here](https://github.com/DataDog/dd-trace-java/blob/8c20fa58d733f6caeeee632cbe00e562018af2f1/dd-java-agent/instrumentation/hazelcast-3.9/src/main/java/datadog/trace/instrumentation/hazelcast39/ClientMessageInstrumentation.java#L69). Please note that you have to declare the context stores that you want to use in the instrumentations.

**Review by bantonsson — COMMENTED** [review]
Thanks for the contribution @monosoul. I don't know enough about Kotlin Coroutines, but it feels like the jobs should already inherit the same `CoroutineContext`, and only if there is a _new_ job should we add a new `CoroutineContext`. Maybe we could just keep track of the associated job in the `CouroutineContext`?

**Comment by monosoul** [comment]
> but it feels like the jobs should already inherit the same `CoroutineContext`, and only if there is a _new_ job should we add a new `CoroutineContext`. Maybe we could just keep track of the associated job in the CouroutineContext?

By inheritance here I mean that each child job (a new job created within another job) inherits it's context. This is the default behavior in coroutines that we were overriding before to make sure each coroutine has it's own instance of scope stack. And this approach is totally fine in general, but unfortunately [this check here](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/common/src/flow/internal/SafeCollector.common.kt#L29) will make the flow collector [throw an exception](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/common/src/flow/internal/SafeCollector.common.kt#L67) here. I suppose JetBrains added that check there expecting scope context can only be modified manually and it kinda makes sense there, but not in our case. So basically now any coroutine will share scope state context element with all it's child jobs, while still having a unique scope state for every job.

File: dd-java-agent/instrumentation/kotlin-coroutines/src/main/java/datadog/trace/instrumentation/kotlin/coroutines/ScopeStateCoroutineContext.java (hunk 1 above)
Comment (monosoul) [reply dd-java-agent/instrumentation/kotlin-coroutines/src/main/java/datadog/trace/instrumentation/kotlin/coroutines/ScopeStateCoroutineContext.java:20]: I don't think I can use it :thinking: It should only be called from instrumentation, but I need to access this map from `ScopeStateCoroutineContext`

Comment (bantonsson) [reply dd-java-agent/instrumentation/kotlin-coroutines/src/main/java/datadog/trace/instrumentation/kotlin/coroutines/ScopeStateCoroutineContext.java:20]: That is only partially true. You can only get access to a generated `ContextStore` in an aspect, but it's perfectly fine to store it in a helper class and use it there. Example [here](https://github.com/DataDog/dd-trace-java/blob/acb275906752bc220630dc268f7dcede14c68805/dd-java-agent/instrumentation/netty-3.8/src/main/java/datadog/trace/instrumentation/netty38/client/HttpClientResponseTracingHandler.java#L20)

Comment (monosoul) [reply dd-java-agent/instrumentation/kotlin-coroutines/src/main/java/datadog/trace/instrumentation/kotlin/coroutines/ScopeStateCoroutineContext.java:20]: Hmm, I tried that but it was giving me runtime exceptions :thinking: Okay, let me try again, maybe I missed something

Comment (monosoul) [reply dd-java-agent/instrumentation/kotlin-coroutines/src/main/java/datadog/trace/instrumentation/kotlin/coroutines/ScopeStateCoroutineContext.java:20]: Yep, it works, probably I missed something last time. Thanks!

Comment (monosoul) [reply dd-java-agent/instrumentation/kotlin-coroutines/src/main/java/datadog/trace/instrumentation/kotlin/coroutines/ScopeStateCoroutineContext.java]: Done

**Review by bantonsson — APPROVED** [review]
